### ISSUE TITLE
Model resetting every 600 iterations?

### ISSUE THREAD

**Issue by harpone** [PyTorchLightning/pytorch-lightning#3348]
## 🐛 Bug
I'm getting a weird model param reset (maybe, not sure) every 600 iterations and haven't been able to pinpoint the reason... happens both on 2xV100 GPUs and v3-8 TPUs.

![Screenshot from 2020-09-04 14-19-26](https://user-images.githubusercontent.com/5112840/92233947-db126e00-eeb9-11ea-8c1f-7c3cd64db155.png)

Does anyone happen to know if something mystic is taking place at iteration 600? I'm using fairly default args and `trainer = Trainer.from_argparse_args(args)`.

### Environment

Happens both on PTL 0.9.0 and 0.9.1rc1. Probably not an environment issue, since the CUDA and XLA environments are quite different... I'm hoping someone will recognize the magical number 600.

**Comment by harpone** [comment]
Update: actually this seems to be due to Weights and Biases logging! I'm looking at the stdout loss and it's not blowing up... investigating...

**Comment by teddykoker** [comment]
This seems to not be an issue with lightning, closing for now.

**Comment by harpone** [comment]
OK it's actually not wandb, since I'm getting this with tensorboard as well...

**Comment by teddykoker** [comment]
Could you try to verify with a different model? Or potentially share your code? We want to make sure this isn’t some weird behavior with your particular model.

**Comment by harpone** [comment]
Can't share code unfortunately because there's lots of dependencies and it's mixed with other company code... but I made some progress. Turns out the number of dataloader workers screw things up really bad for all logging basically (at least W&B and Tensorboard). Here's some experiments with 1, 21, and 24 workers (I was using 24 workers before, which resulted in the "reset" at iter=600, but really that's just 24 * 25 so I think 25 is the "magic number" here):

![Screenshot from 2020-09-08 12-35-46](https://user-images.githubusercontent.com/5112840/92460298-79f7dc80-f1d0-11ea-92b4-547385d9f5a8.png)

So all is OK with 1 worker, but clearly there's some weird pseudo-random logging going on that depends on the workers somehow...

I was logging the above accuracies by calling `self.trainer.logger.experiment.log({'global_step': self.global_step, 'loss_xent_manual': loss_xent, 'acc_manual': acc})`, but similarly weird stuff happens with the `Result` classes. 

I'm doing more or less this at train step:
```
    def training_step(self, batch, batch_idx):

        xs, ys = batch

        _, heads_out = self(xs)

        loss_dct = self.compute_losses(heads_out, ys)
        loss = loss_dct['loss']

        result = pl.TrainResult(minimize=loss)
        result.log('loss_trn', loss_dct['loss'], sync_dist=True)
        if self.args.get('classifier_head', False):
            result.log('loss_xent_trn', loss_dct['loss_xent'], sync_dist=True)
            result.log('acc_trn', loss_dct['acc'], sync_dist=True)
        if self.args.get('conv_head', False):
            result.log('loss_sim_trn', loss_dct['loss_sim'], sync_dist=True)

        if 1:

            # manual loss_xent and acc:
            targets = ys['class_id'].to(dtype=torch.int64)
            preds = torch.argmax(heads_out['classifier_head'], dim=-1).to(targets.dtype)
            acc = torch.eq(targets, preds).float().mean().item()
            loss_xent = self.loss_class_fn(heads_out['classifier_head'], targets).item()

            self.trainer.logger.experiment.log({'global_step': self.global_step,
                                                'loss_xent_manual': loss_xent,
                                                'acc_manual': acc})

        return result
```

and a similar validation step.

I'm using `distributed_backend = 'ddp'` in the Trainer args.

I'll go through an entire train/log step line by line with a debugger.

**Comment by harpone** [comment]
Actually it's not the logging - this happens at every `training_step` even if I just print the `acc` above... what's going on? Can't I use more than 1 dataloader worker??

**Comment by harpone** [comment]
OK it was actually this issue: https://github.com/pytorch/pytorch/issues/5059

So I had 24 dataloaders, each seeded exactly the same way => I got 24 identical minibatches, and the model quickly overfit to those!

I haven't actually used seeds in training at all... do I really need to `seed_everything`?

**Comment by teddykoker** [comment]
Are you using randomly generated data? DDP should split your data among the accelerators if it is a standard pytorch `Dataset`. It is very important to seed everything when using DDP as each accelerator will initialize it's own model, and they should have the same random initialization to work best.

**Comment by harpone** [comment]
Yeah I am, but there's 24 dataloader workers *per each accelerator*, which are all initialized with the same seed...

BTW this is potentially a *very* dangerous bug - I noticed it only because I was using such a high number of workers. If one used e.g. 4 or less, it wouldn't really show up at all, except probably as somewhat deteriorated training performance...

So I really suggest some safeguards when using `seed_everything`.

**Comment by harpone** [comment]
Wouldn't it be better to do one `all_reduce` step before training starts to make sure all accelerators have models with same weights? I wouldn't really want to `seed_everything` in the first place...

**Comment by harpone** [comment]
or something like checking if the `worker_init_fn` is set when using `seed_everything`, like

```
loader = torch.utils.data.DataLoader(...)
if loader.worker_init_fn is None:
    (log a warning)
```

**Comment by harpone** [comment]
hmm yeah that's right... but I was using 'ddp', not 'ddp_spawn'... I guess 'ddp' uses the fork method to distribute the training, so the seed seems to be shared across the dataloader workers if you don't use a `worker_init_fn` that will explicitly. 

Anyway, adding the `worker_init_fn` to do the seed increment explicitly definitely fixes the issue completely.
